IPython/core/display.py

```diff
@@ -139,6 +141,117 @@ def _new_id():
     """Generate a new random text id with urandom"""
     return b2a_hex(os.urandom(16)).decode('ascii')
 
+class RecursiveObject:
+    """
+    Default recursive object that provides a recursion repr if needed.
+
+    You may register a formatter for this object that will be call when
+    recursion is reached.
+    """
+
+    def __init__(self, already_seen):
+        pass
+
+    def __repr__(self):
+        return '<recursion ...>'
+
+    def _repr_html_(self):
+        return '&lt;recursion ...&gt;'
+
+
+class DataMetadata(tuple):
+
+    @property
+    def data(self):
+        return self[0]
```

Comment:
I had issue with namedtuple, where it keep saying in soma place the `.data` was not defined but `[0]` was working. I'm confused as to exactly why but I can retry.